### PR TITLE
url/url: Complete implementation of Origin

### PR DESCRIPTION
**Pull request by Zer0-One**

### CODE REVIEW

File: url/url.h [hunk 1]
```diff
@@ -28,15 +28,61 @@ struct Host {
     std::variant<std::string, std::uint32_t, std::array<std::uint16_t, 8>> data;
 
     std::string serialize() const;
+
+    bool operator==(Host const &b) const {
```
Comment (robinlinden) [inline url/url.h:32]: Looks like this can be defaulted?

File: url/url.h [hunk 2]
```diff
@@ -28,15 +28,61 @@ struct Host {
...
     Host host;
     std::optional<std::uint16_t> port;
     std::optional<std::string> domain;
-    // Need this placeholder until I figure out what "opaqueness" means for an origin in this context
-    bool opaque;
+    // If opaque, then this Origin should serialize to (null). All opaque origins are equal to each other, and not equal
+    // to all non-opaque origins.
+    bool opaque = false;
+
+    std::string serialize();
```
Comment (robinlinden) [inline url/url.h:50]: This and `effective_domain` should probably both be `const`?

**Review by robinlinden — CHANGES_REQUESTED** [review]

File: url/url.h (hunk 1 above)
Comment (Zer0-One) [reply url/url.h:32]: I thought so initially too, but it fails to compile.

```
./url/url.h: In member function 'bool url::Origin::operator==(const url::Origin&) const':
./url/url.h:60:44: error: no match for 'operator==' (operand types are 'const url::Host' and 'const url::Host')
   60 |             if (scheme == b.scheme && host == b.host && port == b.port) {
      |                                       ~~~~ ^~ ~~~~~~
      |                                       |         |
      |                                       |         const url::Host
      |                                       const url::Host
```

Comment (Zer0-One) [reply url/url.h:32]: nevermind, I didn't know this was a thing: https://en.cppreference.com/w/cpp/language/default_comparisons

File: url/url.h [hunk 3]
```diff
@@ -28,15 +28,55 @@ struct Host {
     std::variant<std::string, std::uint32_t, std::array<std::uint16_t, 8>> data;
 
     std::string serialize() const;
+
+    bool operator==(Host const &b) const = default;
```
Comment (robinlinden) [inline url/url.h:32]: No need to name the argument here.

**Review by robinlinden — CHANGES_REQUESTED** [review]
Like in the last PR, you can replace `url/url:` in the commit message w/ `url:`.

One more test is needed to make CI (and me) happy. :P

**Review by robinlinden — APPROVED** [review]

**Pull request merged**